File: sdk/kotlin/src/main/kotlin/com/seocheon/sdk/internal/crypto/PrivateKey.kt
package com.seocheon.sdk.internal.crypto

import org.bouncycastle.crypto.params.ECPrivateKeyParameters
import org.bouncycastle.crypto.params.ECPublicKeyParameters
import org.bouncycastle.crypto.signers.ECDSASigner
import org.bouncycastle.crypto.signers.HMacDSAKCalculator
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.jce.ECNamedCurveTable
import org.bouncycastle.jce.spec.ECNamedCurveParameterSpec
import org.bouncycastle.math.ec.ECPoint
import java.math.BigInteger
import java.security.MessageDigest

/**
 * secp256k1 private key wrapper using BouncyCastle.
 */
class PrivateKey private constructor(
    private val keyBytes: ByteArray,
    private val spec: ECNamedCurveParameterSpec,
) {
    private val d: BigInteger = BigInteger(1, keyBytes)

    /**
     * Returns the compressed 33-byte public key.
     */
    fun pubKey(): ByteArray {
        val q: ECPoint = spec.g.multiply(d).normalize()
        return q.getEncoded(true)
    }

    /**
     * Returns the raw 32-byte private key.
     */
    fun bytes(): ByteArray = keyBytes.copyOf()

    /**
     * Signs the given data using secp256k1. Hashes with SHA-256 first,
     * returns a 64-byte compact signature (R || S).
     */
    fun sign(data: ByteArray): ByteArray {
        val hash = MessageDigest.getInstance("SHA-256").digest(data)

        val signer = ECDSASigner(HMacDSAKCalculator(SHA256Digest()))
        val privParams = ECPrivateKeyParameters(d, org.bouncycastle.crypto.params.ECDomainParameters(
            spec.curve, spec.g, spec.n, spec.h
        ))
        signer.init(true, privParams)

        val components = signer.generateSignature(hash)
        var r = components[0]
        var s = components[1]

        // Enforce low-S (BIP-62)
        val halfN = spec.n.shiftRight(1)
        if (s > halfN) {
            s = spec.n.subtract(s)
        }

        val rBytes = bigIntTo32Bytes(r)
        val sBytes = bigIntTo32Bytes(s)
        return rBytes + sBytes
    }

    companion object {
        private val SPEC: ECNamedCurveParameterSpec = ECNamedCurveTable.getParameterSpec("secp256k1")

        /**
         * Creates a PrivateKey from raw 32-byte private key.
         */
        fun fromBytes(privKeyBytes: ByteArray): PrivateKey {
            require(privKeyBytes.size == 32) {
                "private key must be 32 bytes, got ${privKeyBytes.size}"
            }
            return PrivateKey(privKeyBytes.copyOf(), SPEC)
        }

        /**
         * Verifies a 64-byte compact signature against data using the public key.
         */
        fun verify(pubKeyBytes: ByteArray, data: ByteArray, sigBytes: ByteArray): Boolean {
            if (sigBytes.size != 64) return false

            val hash = MessageDigest.getInstance("SHA-256").digest(data)

            val point = SPEC.curve.decodePoint(pubKeyBytes)
            val pubParams = ECPublicKeyParameters(point, org.bouncycastle.crypto.params.ECDomainParameters(
                SPEC.curve, SPEC.g, SPEC.n, SPEC.h
            ))

            val r = BigInteger(1, sigBytes.sliceArray(0 until 32))
            val s = BigInteger(1, sigBytes.sliceArray(32 until 64))

            val verifier = ECDSASigner()
            verifier.init(false, pubParams)
            return verifier.verifySignature(hash, r, s)
        }

        private fun bigIntTo32Bytes(value: BigInteger): ByteArray {
            val bytes = value.toByteArray()
            return when {
                bytes.size == 32 -> bytes
                bytes.size > 32 -> bytes.sliceArray(bytes.size - 32 until bytes.size)
                else -> ByteArray(32 - bytes.size) + bytes
            }
        }
    }
}

File: sdk/csharp/src/Internal/Crypto/PrivateKey.cs
using System.Security.Cryptography;
using Org.BouncyCastle.Asn1.Sec;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;
using Org.BouncyCastle.Math;
using Seocheon.Sdk.Errors;

namespace Seocheon.Sdk.Internal.Crypto;

/// <summary>
/// secp256k1 private key operations using BouncyCastle.
/// </summary>
public sealed class PrivateKey
{
    private static readonly Org.BouncyCastle.Asn1.X9.X9ECParameters Curve =
        SecNamedCurves.GetByName("secp256k1");

    private static readonly ECDomainParameters Domain =
        new(Curve.Curve, Curve.G, Curve.N, Curve.H);

    private readonly ECPrivateKeyParameters _privateKey;

    private PrivateKey(byte[] keyBytes)
    {
        var d = new BigInteger(1, keyBytes);
        _privateKey = new ECPrivateKeyParameters(d, Domain);
    }

    /// <summary>
    /// Creates a PrivateKey from raw 32-byte key material.
    /// </summary>
    public static PrivateKey FromBytes(byte[] keyBytes)
    {
        if (keyBytes.Length != 32)
            throw SdkErrors.SigningFailed("Private key must be 32 bytes");
        return new PrivateKey(keyBytes);
    }

    /// <summary>
    /// Generates a new random private key.
    /// </summary>
    public static PrivateKey Generate()
    {
        var keyBytes = new byte[32];
        RandomNumberGenerator.Fill(keyBytes);
        return new PrivateKey(keyBytes);
    }

    /// <summary>
    /// Signs data with this private key (deterministic RFC 6979).
    /// Returns a 64-byte compact signature (r || s).
    /// </summary>
    public byte[] Sign(byte[] data)
    {
        var hash = SHA256.HashData(data);
        var signer = new ECDsaSigner(new Org.BouncyCastle.Crypto.Signers.HMacDsaKCalculator(
            new Org.BouncyCastle.Crypto.Digests.Sha256Digest()));
        signer.Init(true, _privateKey);

        var components = signer.GenerateSignature(hash);
        var r = components[0];
        var s = components[1];

        // Ensure low-S (BIP-62)
        var halfOrder = Domain.N.ShiftRight(1);
        if (s.CompareTo(halfOrder) > 0)
            s = Domain.N.Subtract(s);

        var rBytes = BigIntTo32Bytes(r);
        var sBytes = BigIntTo32Bytes(s);

        var signature = new byte[64];
        Buffer.BlockCopy(rBytes, 0, signature, 0, 32);
        Buffer.BlockCopy(sBytes, 0, signature, 32, 32);
        return signature;
    }

    /// <summary>
    /// Returns the compressed public key (33 bytes).
    /// </summary>
    public byte[] GetPubKey()
    {
        var q = Domain.G.Multiply(_privateKey.D).Normalize();
        return q.GetEncoded(true);
    }

    /// <summary>
    /// Returns the raw 32-byte private key.
    /// </summary>
    public byte[] ToBytes()
    {
        return BigIntTo32Bytes(_privateKey.D);
    }

    private static byte[] BigIntTo32Bytes(BigInteger value)
    {
        var bytes = value.ToByteArrayUnsigned();
        if (bytes.Length == 32) return bytes;
        if (bytes.Length > 32) return bytes[^32..];

        var padded = new byte[32];
        Buffer.BlockCopy(bytes, 0, padded, 32 - bytes.Length, bytes.Length);
        return padded;
    }
}
